engine/storage/src/main/java/org/apache/cloudstack/storage/volume/datastore/PrimaryDataStoreHelper.java

```diff
@@ -265,4 +268,45 @@ public boolean deletePrimaryDataStore(DataStore store) {
         return true;
     }
 
-}
+    public void switchToZone(DataStore store) {
+        StoragePoolVO pool = dataStoreDao.findById(store.getId());
+        CapacityVO capacity = _capacityDao.findByHostIdType(store.getId(), Capacity.CAPACITY_TYPE_STORAGE_ALLOCATED);
+        Transaction.execute(new TransactionCallbackNoReturn() {
+            public void doInTransactionWithoutResult(TransactionStatus status) {
+                pool.setScope(ScopeType.ZONE);
+                pool.setPodId(null);
+                pool.setClusterId(null);
+                dataStoreDao.update(pool.getId(), pool);
```

Comment:
`pool.setHypervisor(xxx)` ?
this is important for zone-wide pools

Reply:
The `storage_pool table` will already have the `hypervisor` field set.
(It is set for cluster-wide pools also)

Reply:
@abh1sar 
it is set in recent releases.
However, in some older environments (earlier than #4307 ) , the hypervisor type is NULL for cluster-wide pools.

Reply:
Will add

Reply:
I realized was getting the hypervisor type for cluster scope primary storage from the storage_pool table, which might not have the hypervisor field set.
So have changed that code too. Now I am getting the hypervisor type from clusterVO.